### PR TITLE
rfqmsg+rfq: add execution policy to rfq negotiation

### LINKED ISSUE #2004: [feature]: RFQ: First-Class Limit Orders In Quote Negotiation
## Summary
Today, RFQ orders are mostly constrained by:
- Requested max size (`asset_max_amt` or `payment_max_amt`)
- Quote expiry
- Channel availability and balance shape
- Price-oracle tolerance checks

This is sufficient for basic RFQ flows, but it is not a full limit-order model. We should add explicit limit-order semantics (price bounds, fill constraints, and execution policy) to quote negotiation.

## Why This Matters
Limit-order semantics enable:
- Price protection: avoid execution outside user-defined acceptable price bounds.
- Slippage control: reduce risk from stale hints, spread moves, and short-term volatility.
- Deterministic UX: users can express explicit execution intent instead of best-effort behavior.
- Safer automation: wallet agents can auto-negotiate with tighter risk controls.
- Better policy explainability: clearer machine-readable reject reasons.
- Better multi-peer execution: easier to split across peers while preserving user constraints.

## Terminology
- `IOC` (`Immediate-Or-Cancel`): execute immediately for any amount that satisfies constraints; cancel any unfilled remainder.
- `FOK` (`Fill-Or-Kill`): execute immediately and fully, or cancel the entire order.

### What RFQ can express today
- Buy flow: max asset amount (`AssetMaxAmt`) + expiry + peer + metadata.
- Sell flow: max payment amount (`PaymentMaxAmt`) + expiry + peer + metadata.
- Request/accept negotiation primarily settles on a rate + expiry.

### What is enforced in practice
- Requester-side quote validation checks include quote expiry bounds and oracle tolerance against the accepted rate.
- HTLC/policy enforcement is max-cap based, where sale path policy derives from `AssetMaxAmt` and purchase path policy derives from `PaymentMaxAmt`.
- Liquidity constraints are effectively channel-driven: channel existence checks gate order submission, and channel balances influence selection/sorting.

### Existing but underused protocol surface
- Wire format already includes optional `MinInAsset` and `MinOutAsset` fields.
- These min fields are not first-class in typed request structs and RPC order APIs.

### Current gap
Current behavior is closer to:
- "Market quote with max-size cap + expiry"
not:
- "Limit order with explicit price/fill constraints"

Users cannot directly encode:
- Minimum acceptable price / rate.
- Maximum acceptable price / rate.
- Minimum fill size.
- Execution policy (`IOC` vs `FOK`).

### Workstream A: Add explicit limit constraints to request surfaces
- Add optional limit-price/rate bound fields to buy/sell order APIs.
- Add optional fill bound fields (`min_in_asset`, `min_out_asset`) to typed requests and RPC.
- Keep current max fields as-is for compatibility.
- Define a canonical rate-bound representation that avoids buy/sell direction ambiguity.

### Workstream B: Enforce limits during negotiation
- Responder side (`ResolveRequest`): reject when request limits cannot be satisfied.
- Requester side (`VerifyAcceptQuote`): enforce explicit requester limits in addition to oracle tolerance.
- Return explicit reject reasons for limit failures (price bound miss, min fill miss, insufficient liquidity).

### Workstream C: Add execution policy semantics
- Support explicit execution policy on orders.
- `IOC`: allow partial immediate execution above minimum fill.
- `FOK`: require full immediate execution or reject.
- Ensure policy is respected consistently in quote acceptance and HTLC policy installation.

### Workstream D: Negotiate explicit fill quantities
- Add accepted fill quantity semantics to quote acceptance path.
- Bind policy persistence and HTLC enforcement to negotiated fill values, not only request max caps.
- This removes ambiguity around partial fills and improves deterministic execution.

### Workstream E: Multi-peer selection that respects limits
- When multiple peers are available, rank and split using effective rate and usable capacity.
- Ensure every shard respects the same order limits and execution policy.

## Non-Goals
- Building a centralized order book.
- Long-lived resting/GTC orders in initial implementation.
- Redesigning the external price-oracle architecture.

## Compatibility Expectations
- New fields should be additive and optional.
- Old peers should continue to interoperate under current behavior.
- Limit-specific validation should only activate when limit fields are present.
- Use versioning only where strict incompatibility requires it.

## Acceptance Criteria
- Users can specify explicit limit-price bounds for buy and sell RFQ orders.
- Users can specify minimum fill constraints.
- Users can choose execution policy (`IOC` or `FOK`).
- Quotes violating constraints are rejected with clear machine-readable reasons.
- Accepted quotes cannot pass validation if they violate requester limits.
- Policy/HTLC enforcement is tied to negotiated constraints, not only broad request max caps.
- Test coverage includes buy and sell flows.
- Test coverage includes `IOC` and `FOK` behaviors.
- Test coverage includes backward compatibility with old peers.
- Test coverage includes volatility and timing edge cases.

## Open Questions
- What is the canonical limit-price representation to avoid buy/sell interpretation errors?
- Should both min and max rate bounds be supported initially, or only one directional bound per order type?
- Do we extend `QuoteRespStatus`, add new reject codes, or both?
- Is explicit fill quantity required in the first delivery, or can it follow once request-side limits are live?
- Should passive offer filtering and limit-order checks be merged into one decision pipeline?

### PR DESCRIPTION
**Pull request by jtobin**
Depends on #2048. Partially resolves #2004 (most, though not all, of workstream C).

Adds execution policy semantics (immediate-or-cancel, fill-or-kill) to the RFQ wire protocol and RPC surface, which again, as in #2048, are here enforced at the quote acceptance level. This PR sets up the wire encoding, RPC surface, and validation plumbing for the feature, but nontrivial enforcement is deferred to #2050 (where we have fill quantities to work with).

Feature TLDR:

- Users can choose either an IOC or FOK policy (the former accepts partial fills, the latter requires a complete fill)
- Quotes violating FOK are rejected with a machine-readable reason (FOK_NOT_VIABLE).
- Backward-compatible: field is optional, default is IOC.

Like the functionality added in #2048, the execution policy field is additive and optional; old peers that ignore the new TLV record continue to interoperate.

Here's an illustrative use of the feature, from an e2e regtest demo:

```
Frank buys 1,000 USDX with Fill-or-Kill. The full amount converts
cleanly at the current rate — quote accepted, payment executes.

Market rate: $66,805.00/BTC
✓ Accepted
  RFQ ID: wMzJgyYBcWwK...
  SCID:   17614968261059438554
  Rate:   $66,805.00/BTC
  Amount: 1000 USDX
  Policy: EXECUTION_POLICY_FOK

→ AddInvoice (rfq_id: wMzJgyYBcWwK...)
  1000 USDX from Frank, reusing negotiated quote
  Quote confirmed: wMzJgyYBcWwK...
  Invoice: lnbcrt149689390p1p5ulzhnpp5zkkknghqfhr2l...

→ Payment: Alice → Eve → Frank
  ✓ Payment succeeded
```

### CODE REVIEW

File: rfq/portfolio_pilot.go [hunk 1]
```diff
@@ -443,3 +461,145 @@ func (p *InternalPortfolioPilot) QueryAssetRates(ctx context.Context,
...
+		if !isFOK(r.ExecutionPolicy) {
+			return ValidAcceptQuoteRespStatus
+		}
+
+		units := rfqmath.FixedPoint[rfqmath.BigInt]{
+			Coefficient: rfqmath.NewBigIntFromUint64(
+				r.AssetMaxAmt,
+			),
+			Scale: 0,
+		}
```
[the remark this answers is not in the document]
Comment (jtobin) [reply rfq/portfolio_pilot.go]: Fixed in 5185d1608d95711e210aa27b1373f12616f6f379.

File: rfq/portfolio_pilot.go [hunk 2]
```diff
@@ -443,3 +461,145 @@ func (p *InternalPortfolioPilot) QueryAssetRates(ctx context.Context,
...
+				units := rfqmath.FixedPoint[rfqmath.BigInt]{
+					Coefficient: rfqmath.NewBigIntFromUint64(
+						minAmt,
+					),
+					Scale: 0,
+				}
+				msat := rfqmath.UnitsToMilliSatoshi(
+					units, acceptedRate,
+				)
+				return msat == 0
```
[the remark this answers is not in the document]
Comment (jtobin) [reply rfq/portfolio_pilot.go]: Fixed in 5185d1608d95711e210aa27b1373f12616f6f379.

File: rfqmsg/buy_request.go [hunk 3]
```diff
@@ -153,13 +182,41 @@ func NewBuyRequestFromWire(wireMsg WireMessage,
 		},
 	)
 
+	// Extract optional min asset amount.
+	var assetMinAmt fn.Option[uint64]
+	msgData.MinInAsset.WhenSome(
+		func(r tlv.RecordT[tlv.TlvType23, uint64]) {
+			assetMinAmt = fn.Some(r.Val)
+		},
+	)
```
[the remark this answers is not in the document]
Comment (jtobin) [reply rfqmsg/buy_request.go:193]: Fixed in 5185d1608d95711e210aa27b1373f12616f6f379.

File: rfqmsg/sell_request.go [hunk 4]
```diff
@@ -147,6 +177,33 @@ func NewSellRequestFromWire(wireMsg WireMessage,
...
 
+	// Extract optional min payment amount.
+	var paymentMinAmt fn.Option[lnwire.MilliSatoshi]
+	msgData.MinOutAsset.WhenSome(
+		func(r tlv.RecordT[tlv.TlvType25, uint64]) {
+			paymentMinAmt = fn.Some(
+				lnwire.MilliSatoshi(r.Val),
+			)
+		},
+	)
```
[the remark this answers is not in the document]
Comment (jtobin) [reply rfqmsg/sell_request.go:190]: Fixed in 5185d1608d95711e210aa27b1373f12616f6f379.

File: rfq/portfolio_pilot.go [hunk 5]
```diff
@@ -443,3 +461,145 @@ func (p *InternalPortfolioPilot) QueryAssetRates(ctx context.Context,
...
+
+	case *rfqmsg.SellRequest:
+		notMet := fn.MapOptionZ(
+			r.PaymentMinAmt,
+			func(minAmt lnwire.MilliSatoshi) bool {
+				units := rfqmath.MilliSatoshiToUnits(
+					minAmt, acceptedRate,
+				)
+				zero := rfqmath.NewBigIntFromUint64(0)
+				return units.Coefficient.Equals(zero)
```
[the remark this answers is not in the document]
Comment (jtobin) [reply rfq/portfolio_pilot.go:586]: Fixed in 5185d1608d95711e210aa27b1373f12616f6f379.

File: rfq/portfolio_pilot_rpc.go
Comment (darioAnongba) [inline rfq/portfolio_pilot_rpc.go:458]: Missing execution policy for external PP?

Comment (darioAnongba) [inline rfq/portfolio_pilot_rpc.go:513]: Missing execution policy for external PP? The ideal here is an integration test checking that the requests to external RPCs match what we expect. Can the `InternalPP` be wrapped easily into an `RpcPP`?

Comment (jtobin) [reply rfq/portfolio_pilot_rpc.go:458]: These are added later in the series (in #2056) but should probably be ported here.

Comment (jtobin) [reply rfq/portfolio_pilot_rpc.go:458]: Resolved in latest changeset.

Comment (jtobin) [reply rfq/portfolio_pilot_rpc.go:513]: Resolved in latest changeset. I added unit tests, rather than an itest, as they seem sufficient to test request equality.

File: rfqmsg/buy_request.go [hunk 6]
```diff
@@ -85,7 +90,8 @@ func NewBuyRequest(peer route.Vertex, assetSpecifier asset.Specifier,
 	assetMaxAmt uint64, assetMinAmt fn.Option[uint64],
 	assetRateLimit fn.Option[rfqmath.BigIntFixedPoint],
 	assetRateHint fn.Option[AssetRate],
-	oracleMetadata string) (*BuyRequest, error) {
+	oracleMetadata string,
+	execPolicy fn.Option[ExecutionPolicy]) (*BuyRequest, error) {
```
Comment (GeorgeTsagk) [inline rfqmsg/buy_request.go:94]: nit: this commit won't compile, callers are not updated accordingly

Comment (jtobin) [reply rfqmsg/buy_request.go:94]: I refute it thusly! 😁 

```
$ git checkout fdb955486 && make build > /dev/null 2>&1 && echo "built"
HEAD is now at fdb95548 rfqmsg: add execution policy type and wire encoding
built
```
